lib/Dialect/Arc/Transforms/LowerStateRewrite.cpp

```diff
@@ -0,0 +1,1206 @@
+  if (initial)
+    return success();
+  if (llvm::is_contained(operands, Value{}))
+    return failure();
+
+  // Expose the `seq.initial` operands as values for the block arguments.
+  for (auto [arg, operand] : llvm::zip(op.getBody().getArguments(), operands))
+    module.loweredValues[{arg, Phase::Initial}] = operand;
+
+  // Lower each op in the body.
+  for (auto &bodyOp : op.getOps()) {
+    if (isa<seq::YieldOp>(bodyOp))
+      continue;
+
+    // Clone the operation.
+    auto *clonedOp = module.initialBuilder.clone(bodyOp);
+    auto result = clonedOp->walk([&](Operation *nestedClonedOp) {
+      for (auto &operand : nestedClonedOp->getOpOperands()) {
+        if (clonedOp->isAncestor(operand.get().getParentBlock()->getParentOp()))
+          continue;
+        auto value = module.requireLoweredValue(operand.get(), Phase::Initial,
+                                                nestedClonedOp->getLoc());
+        if (!value)
+          return WalkResult::interrupt();
+        operand.set(value);
+      }
+      return WalkResult::advance();
+    });
+    if (result.wasInterrupted())
+      return failure();
+
+    // Keep track of the results.
+    for (auto [result, lowered] :
+         llvm::zip(bodyOp.getResults(), clonedOp->getResults()))
+      module.loweredValues[{result, Phase::Initial}] = lowered;
+  }
+
+  // Expose the operands of `seq.yield` as results from the initial op.
+  auto *terminator = op.getBodyBlock()->getTerminator();
+  for (auto [result, operand] :
+       llvm::zip(op.getResults(), terminator->getOperands())) {
+    auto value = module.requireLoweredValue(operand, Phase::Initial,
+                                            terminator->getLoc());
+    if (!value)
+      return failure();
+    module.loweredValues[{result, Phase::Initial}] = value;
+  }
+
+  return success();
+}
+
+/// Lower `llhd.final` ops into `scf.execute_region` ops in the `arc.final` op.
+LogicalResult OpLowering::lower(llhd::FinalOp op) {
+  if (phase != Phase::Final)
+    return op.emitOpError() << "cannot be lowered in the " << phase << " phase";
+
+  // Determine the uses of values defined outside the op.
+  SmallVector<Value> externalOperands;
+  op.walk([&](Operation *nestedOp) {
+    for (auto value : nestedOp->getOperands())
+      if (!op->isAncestor(value.getParentBlock()->getParentOp()))
+        externalOperands.push_back(value);
+  });
+
+  // Make sure that all uses of external values are lowered first.
+  IRMapping mapping;
+  for (auto operand : externalOperands) {
+    auto lowered = lowerValue(operand, Phase::Final);
+    if (!initial && !lowered)
+      return failure();
+    mapping.map(operand, lowered);
+  }
+  if (initial)
+    return success();
+
+  // Handle the simple case where the final op contains only one block, which we
+  // can inline directly.
+  if (op.getBody().hasOneBlock()) {
+    for (auto &bodyOp : op.getBody().front().without_terminator())
+      module.finalBuilder.clone(bodyOp, mapping);
```

Comment:
If we make this pass operate on OpInterfaces in the future, we probably also want to call `lower` on these nested operations.

Reply:
Yeah this needs some careful thought to figure out how exactly ops with nested regions would be lowered.